### PR TITLE
concurrent tar cache file safe

### PR DESCRIPTION
**Pull request by chengjoey**
#### What type of this PR

Add one of the following kinds:
/kind bugfix

#### What this PR does / why we need it:
concurrent tar cache file safe, lock cache path when tar file

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda-org.erda.cloud/erda/dop/projects/387/issues/all?id=222146&issueFilter__urlQuery=eyJzdGF0ZUJlbG9uZ3MiOlsiT1BFTiIsIldPUktJTkciLCJXT05URklYIiwiUkVPUEVOIl0sInN0YXRlcyI6WzQ0MDIsNzEwNCw3MTA1LDQ0MDMsNDQwNCw3MTA2LDQ0MDYsNDQwNyw0NDEyLDQ1MzgsNDQxMyw0NDE0LDQ0MTUsNDQxNl0sImFzc2lnbmVlSURzIjpbIjEwMDEyMDUiXX0%3D&issueTable__urlQuery=eyJwYWdlTm8iOjF9&issueViewGroup__urlQuery=eyJ2YWx1ZSI6ImthbmJhbiIsImNoaWxkcmVuVmFsdWUiOnsia2FuYmFuIjoiZGVhZGxpbmUifX0%3D&iterationID=680&type=BUG)

#### Specified Reviewers:

/assign @your-reviewer

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |Fix the bug that concurrent tar cache file safe              |
| 🇨🇳 中文    |  修复了pipeline cache文件多分支碰撞的问题            |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: modules/actionagent/agenttool/lock.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	"path/filepath"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/erda-project/erda/pkg/filehelper"
+)
+
+const (
+	lockFileName = "erda.lock"
```
Comment (sfwn) [inline modules/actionagent/agenttool/lock.go:29]: cache.lock

File: modules/actionagent/agenttool/lock.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+func WaitingPathUnlock(path string, maxLimit int) error {
+	limitTimer := time.NewTimer(time.Duration(float64(maxLimit) * float64(time.Second)))
+	ticker := time.NewTicker(time.Second)
+	defer limitTimer.Stop()
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-limitTimer.C:
+			return fmt.Errorf("after %d seconds, the path: %s is still not unlocked", maxLimit, path)
```
Comment (sfwn) [inline modules/actionagent/agenttool/lock.go:61]: standardize log, add prefix [cache]

File: modules/actionagent/agenttool/lock.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	lockFile := filepath.Join(path, lockFileName)
+	if err := filehelper.CheckExist(lockFile, false); err != nil {
+		return false
+	}
+	return true
+}
+
+func WaitingPathUnlock(path string, maxLimit int) error {
+	limitTimer := time.NewTimer(time.Duration(float64(maxLimit) * float64(time.Second)))
+	ticker := time.NewTicker(time.Second)
```
Comment (sfwn) [inline modules/actionagent/agenttool/lock.go:54]: second is a too-large step. use millisecond at beginning.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/actionagent/agenttool/lock.go (hunk 1 above)
Comment (chengjoey) [reply modules/actionagent/agenttool/lock.go]: adjusted

File: modules/actionagent/agenttool/lock.go (hunk 2 above)
Comment (chengjoey) [reply modules/actionagent/agenttool/lock.go]: adjusted

File: modules/actionagent/agenttool/lock.go (hunk 3 above)
Comment (chengjoey) [reply modules/actionagent/agenttool/lock.go:59]: change to check locked immediately

Comment (sfwn) [reply modules/actionagent/agenttool/lock.go:59]: second still too-large.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/actionagent/agenttool/lock.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+				checkTimer.Reset(calculateNextCheckTimeDuration(checkedTimes))
+				continue
+			}
+
+			return nil
+		}
+	}
+}
+
+func calculateNextCheckTimeDuration(checkedTimes uint64) time.Duration {
```
Comment (sfwn) [inline modules/actionagent/agenttool/lock.go:84]: add ut for this method pls.

**Review by sfwn — CHANGES_REQUESTED** [review]

**Review by sfwn — APPROVED** [review]

**Pull request merged**